File: 프로그래머스/DFS_타겟넘버/타겟넘버.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;
vector<int> num;
int tar, numsize, answer = 0;

void dfs(int idx, int sum) {
	if (idx == numsize) {
		if (sum == tar) {
			answer++;
		}
		return;
	}
	dfs(idx + 1, sum + num[idx]);
	dfs(idx + 1, sum - num[idx]);
}

int solution(vector<int> numbers, int target) {
	answer = 0;
	tar = target;
	num = numbers;
	numsize = numbers.size();
	dfs(0, 0);
	return answer;
}

File: 프로그래머스/DFS_타겟넘버/타겟넘버.py
answer = 0
def dfs(graph, idx, target, sum):
    global answer
    if idx == len(graph):
        if(sum == target):
            answer+=1
        return True
    dfs(graph, idx+1, target, sum+graph[idx])
    dfs(graph, idx+1, target, sum -graph[idx])
def solution(numbers, target):
    
    dfs(numbers, 0, target, 0)
    
    return answer
